guide/plot-composition.qmd

```diff
@@ -0,0 +1,117 @@
+jupyter: python3
+---
+
+Plot composition is the process of combining multiple plots together.
+Plotnine overloads the `|` and `/` operators for putting plots side-by-side or stacked vertically. It also provides the `&`, `*`, and `+` operators for adding components to parts of composed plots.
+
+This page covers the basics of plot composition. See the [Compose Reference page](`plotnine.composition.Compose`) for more details.
+
+## Big example
+
+```{python}
+# | warning: false
+from plotnine import *
+from plotnine.data import mtcars
+
+p1 = ggplot(mtcars) + geom_point(aes("wt", "mpg")) + labs(tag="a)")
+p2 = ggplot(mtcars) + geom_boxplot(aes("wt", "disp", group="gear")) + labs(tag="b)")
+p3 = ggplot(mtcars) + geom_smooth(aes("disp", "qsec")) + labs(tag="c)")
+p4 = ggplot(mtcars) + geom_bar(aes("carb"))
+
+(p1 | p2 | p3) / p4
+```
+
+Notice the following pieces in the code:
+
+* `|` places plots side-by-side.
+* `/` stacks plots vertically.
+* `labs(tag=...)` adds a label to the top-left of each plot.
+
+## Tag position
+
+Use theme(plot_tag_position=...) to change the position of the tag on a plot.
+
+```{python}
+p1 | (p2 + theme(plot_tag_position="bottom-right"))
+```
+
+Notice that the tag for `p2` ("b)") is now at the bottom-right.
+
+## Adding to composed plots
+
+Use the `&`, `*`, or `+` operators to add components like geoms, scales, or themes to plots in a plot composition. The `&` operator adds to all plots in a composition, while `*` and `+` add to the first or last plot, respectively.
+
+```{python}
+# | eval: false
+composed = p1 | p2
+
+composed & theme_minimal()  # themes both plots
+composed * theme_minimal()  # themes first plot (p1)
+composed + theme_minimal()  # themes last plot (p2)
+```
+
+## Adding margins to plots
+
+Use options like `theme(plot_margin_left=...)` to add margins to plots. Margins can be added to the left, right, top, or bottom of a plot.
+
+```{python}
+p1 | p2 + theme(plot_margin_left=0.20)
+```
+
+Notice that a left margin was added to the second plot (`p2`).
+
+## Inserting space between plots
+
+Use `plot_spacer()` to insert space between plots.
+
+```{python}
+from plotnine.composition import plot_spacer
+
+(p1 | p2) / (p1 | plot_spacer() | p2)
+```
+
+Notice that the top row has the plots side-by-side, while the bottom row puts a space between them.
+
+## Nesting level
+
+Nesting level determines the proportion of space a plot (or group of plots) is given. Essentially, each group of plots is a nesting level, and because groups of plots can hold other groups, you can have multiple levels of nesting. This may sound confusing, so this section will start with an example, before explaining nesting levels in more detail.
+
+Use the `-` operator to put the right-hand plot group twice as much space as each individual plot on the left.
+
+
+```{python}
+# | warning: false
+top = p1 | p2 | p3 | p4  # all on same nesting level
+bot = (p1 | p2 | p3) - p4  # p1, p2, p3 grouped (lower nesting level)
```

Comment:
@has2k1 is it by design that p4 is the width of two of the other plots here? (i.e. p4 seems to be the width roughly of p2 and p3 combined?). Or should it be the width of the lhs group (p1, p2, p3)

Happy with w/e, just want to make sure I described the behavior correctly!

Reply:
https://pr-24--helpful-monstera-4bc9c3.netlify.app/guide/plot-composition#nesting-level

<img width="826" height="634" alt="image" src="https://github.com/user-attachments/assets/c164816f-5b1c-41fd-86ca-4c7ce66fb0ec" />

Reply:
It is the _width of the panels_ and not width of the plots.

So for the bottom one, the width of the panel in `p4` is equal to the sum of the widths of the panels in `p1`, `p2` and `p3`. That means the composition `(p1 | p2 | p3)` is at the same nesting level as the plot `p4`.

Intuitively, one would expect to get such an output using `(p1 | p2 | p3) | p4`, but the operator precedence and left to right evaluation make it impossible. Yet if the grouping is on the right, `p1 | (p2 | p3 | p4)`, then the result and the intuition match!

Reply:
Ah okay -- do you mean something like this? So there are two pieces that affect width, panel width and some other component (where the tag label is)? Does that sound right?

<img width="2800" height="2000" alt="image" src="https://github.com/user-attachments/assets/c6d6977f-5e7c-43f9-ad85-cc86e846a339" />

Reply:
<img width="1167" height="374" alt="Screenshot 2025-07-17 at 19 13 30" src="https://github.com/user-attachments/assets/a8be0747-5762-4961-a3f5-f016bb9564b1" />